pkg/build/builder/cmd/scmauth/password.go

```diff
@@ -0,0 +1,131 @@
+package scmauth
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/openshift/origin/pkg/util/file"
+)
+
+const (
+	DefaultUsername      = "builder"
+	UsernamePasswordName = "password"
+	UsernameSecret       = "username"
+	PasswordSecret       = "password"
+	TokenSecret          = "token"
+	UserPassGitConfig    = `# credential git config
+[credential]
+   helper = store --file=%[1]s
+
+[http]
+   cookieFile = %[2]s
+`
+	TokenCookie = "%[1]s\tTRUE\t/\tFALSE\t0\tGIT_TOKEN\t%[2]s"
+)
+
+// UsernamePassword implements SCMAuth interface for using Username and Password credentials
+type UsernamePassword struct {
+	SourceURL url.URL
+}
+
+// Setup creates a gitconfig fragment that includes a substitution URL with the username/password
+// included in the URL
+func (u UsernamePassword) Setup(baseDir string) error {
+
+	// Only apply to https and http URLs
+	if scheme := strings.ToLower(u.SourceURL.Scheme); scheme != "http" && scheme != "https" {
+		return nil
+	}
+
+	// Determine username
+	// 1. Look for a username secret
+	username, err := readSecret(baseDir, UsernameSecret)
+	if err != nil {
+		return err
+	}
+	// 2. If not provided, look at the username in the URL
+	if username == "" && u.SourceURL.User != nil {
+		username = u.SourceURL.User.Username()
+	}
+	// 3. If still not found, use a default username
+	if username == "" {
+		username = DefaultUsername
+	}
+
+	// Determine password
+	// 1. Look for a token secret
+	password, err := readSecret(baseDir, TokenSecret)
+	if err != nil {
+		return err
+	}
+	// 2. Look for a password secret
+	if password == "" {
+		password, err = readSecret(baseDir, PasswordSecret)
+		if err != nil {
+			return err
+		}
+	}
+	// 3. Look for a password in the URL
+	if password == "" && u.SourceURL.User != nil {
+		password, _ = u.SourceURL.User.Password()
+	}
+
+	gitcredentials, err := ioutil.TempFile("", "gitcredentials.")
+	if err != nil {
+		return err
+	}
+	defer gitcredentials.Close()
+	gitconfig, err := ioutil.TempFile("", "gitcredentialscfg.")
+	if err != nil {
+		return err
+	}
+	defer gitconfig.Close()
+	cookiefile, err := ioutil.TempFile("", "gitcookie.")
+	if err != nil {
+		return err
+	}
+	defer cookiefile.Close()
+
+	usernamePasswordURL := u.SourceURL
+	usernamePasswordURL.User = url.UserPassword(username, password)
+	fmt.Fprintf(gitconfig, UserPassGitConfig, gitcredentials.Name(), cookiefile.Name())
+	fmt.Fprintf(gitcredentials, usernamePasswordURL.String())
+	fmt.Fprintf(cookiefile, TokenCookie, u.SourceURL.Host, password)
```

Comment:
The internal git server is checked in and working.  We need something.
This story isn't useful if it doesn't enable the gitserver to get tested
and working.  So we need something.

On Fri, Sep 4, 2015 at 12:06 PM, Ben Parees <notifications@github.com>
wrote:

> In pkg/build/builder/cmd/scmauth/password.go
> <https://github.com/openshift/origin/pull/4387#discussion_r38767003>:
>
> > +	gitconfig, err := ioutil.TempFile("", "gitcredentialscfg.")
> > +	if err != nil {
> > +		return err
> > +	}
> > +	defer gitconfig.Close()
> > +	cookiefile, err := ioutil.TempFile("", "gitcookie.")
> > +	if err != nil {
> > +		return err
> > +	}
> > +	defer cookiefile.Close()
> > +
> > +	usernamePasswordURL := u.SourceURL
> > +	usernamePasswordURL.User = url.UserPassword(username, password)
> > +	fmt.Fprintf(gitconfig, UserPassGitConfig, gitcredentials.Name(), cookiefile.Name())
> > +	fmt.Fprintf(gitcredentials, usernamePasswordURL.String())
> > +	fmt.Fprintf(cookiefile, TokenCookie, u.SourceURL.Host, password)
>
> i really don't know, but considering we don't even have a use case that
> requires this today (since we don't have an internal git server), it seems
> like we could at least wait and see and put the proxy in if we need it,
> when we go to support internal git w/ openshift auth.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4387/files#r38767003>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift